api/envoy/api/v2/core/address.proto

```diff
@@ -74,6 +74,29 @@ message TcpKeepalive {
   google.protobuf.UInt32Value keepalive_interval = 3;
 }
 
+// Generic socket option message. This would be used to set socket options that
+// might not exist in upstream kernels or precompiled Envoy binaries.
+message SocketOption {
+  // An optional name to give this socket option
+  // No special meaning is assumed.
+  string readable_name = 1;
+  // Corresponding to the level value passed to setsockopt, such as IPPROTO_TCP
+  int32 level = 2;
+  // The numeric name as passed to setsockopt
+  int32 name = 3;
+  // The setsopt option value. Integers should be encoded in network byte order.
+  bytes value = 4;
```

Comment:
Given that many socket options take of value of type int, maybe make this oneof(int, bytes), so people don't have to worry about byte order for the common case.